File: fetch-hours.js
#!/usr/bin/env node
import fetch from 'node-fetch';

const SUPABASE_URL = 'https://jgvckilmltimabfdvaly.supabase.co';
const SUPABASE_KEY = 'sb_publishable_TiO_hACvlsn4mfFazsHTlg_i9BnzCfr';
const GOOGLE_API_KEY = process.env.GOOGLE_PLACES_API_KEY;

if (!GOOGLE_API_KEY) {
  console.error('❌ Error: GOOGLE_PLACES_API_KEY environment variable not set');
  console.error('Please set it with: export GOOGLE_PLACES_API_KEY="your-api-key"');
  process.exit(1);
}

const headers = {
  'apikey': SUPABASE_KEY,
  'Authorization': `Bearer ${SUPABASE_KEY}`,
  'Content-Type': 'application/json',
  'Prefer': 'return=representation'
};

// Fetch all locations from Supabase
async function fetchLocations() {
  try {
    const response = await fetch(`${SUPABASE_URL}/rest/v1/locations?select=*`, { headers });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return await response.json();
  } catch (err) {
    console.error('Failed to fetch locations:', err);
    throw err;
  }
}

// Search for place using Google Places API (Text Search)
async function searchPlace(name, address) {
  try {
    const query = encodeURIComponent(`${name} ${address} Los Angeles`);
    const url = `https://maps.googleapis.com/maps/api/place/textsearch/json?query=${query}&key=${GOOGLE_API_KEY}`;

    const response = await fetch(url);
    const data = await response.json();

    if (data.status === 'OK' && data.results.length > 0) {
      return data.results[0].place_id;
    }
    return null;
  } catch (err) {
    console.error(`Error searching for place ${name}:`, err);
    return null;
  }
}

// Get place details including opening hours
async function getPlaceDetails(placeId) {
  try {
    const url = `https://maps.googleapis.com/maps/api/place/details/json?place_id=${placeId}&fields=opening_hours&key=${GOOGLE_API_KEY}`;

    const response = await fetch(url);
    const data = await response.json();

    if (data.status === 'OK' && data.result.opening_hours) {
      return data.result.opening_hours;
    }
    return null;
  } catch (err) {
    console.error(`Error fetching details for place ${placeId}:`, err);
    return null;
  }
}

// Update location in Supabase with hours data
async function updateLocationHours(locationId, hours) {
  try {
    const response = await fetch(`${SUPABASE_URL}/rest/v1/locations?id=eq.${locationId}`, {
      method: 'PATCH',
      headers,
      body: JSON.stringify({ hours })
    });

    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return true;
  } catch (err) {
    console.error(`Failed to update location ${locationId}:`, err);
    return false;
  }
}

// Main function
async function main() {
  console.log('🚀 Starting business hours fetch for LA Trip Map locations...\n');

  // Fetch all locations
  console.log('📍 Fetching locations from Supabase...');
  const locations = await fetchLocations();
  console.log(`✓ Found ${locations.length} locations\n`);

  let successCount = 0;
  let failCount = 0;
  let skipCount = 0;

  // Process each location
  for (let i = 0; i < locations.length; i++) {
    const location = locations[i];
    console.log(`[${i + 1}/${locations.length}] Processing: ${location.name}`);

    // Check if hours already exist
    if (location.hours) {
      console.log(`  ⏭️  Already has hours data, skipping`);
      skipCount++;
      continue;
    }

    // Search for place
    console.log(`  🔍 Searching Google Places...`);
    const placeId = await searchPlace(location.name, location.address || '');

    if (!placeId) {
      console.log(`  ❌ Could not find place on Google Places`);
      failCount++;
      continue;
    }

    console.log(`  ✓ Found place ID: ${placeId}`);

    // Get place details
    console.log(`  📋 Fetching business hours...`);
    const hours = await getPlaceDetails(placeId);

    if (!hours) {
      console.log(`  ⚠️  No hours data available for this place`);
      failCount++;
      continue;
    }

    // Update Supabase
    console.log(`  💾 Updating database...`);
    const updated = await updateLocationHours(location.id, hours);

    if (updated) {
      console.log(`  ✅ Successfully updated with hours data`);
      successCount++;
    } else {
      console.log(`  ❌ Failed to update database`);
      failCount++;
    }

    // Rate limiting - wait 100ms between requests to avoid hitting API limits
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  // Summary
  console.log('\n📊 Summary:');
  console.log(`  ✅ Success: ${successCount}`);
  console.log(`  ❌ Failed: ${failCount}`);
  console.log(`  ⏭️  Skipped: ${skipCount}`);
  console.log(`  📍 Total: ${locations.length}`);

  if (successCount > 0) {
    console.log('\n🎉 Business hours have been successfully fetched and stored!');
  }
}

// Run the script
main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: fetch_hours.py
#!/usr/bin/env python3
"""
Fetch and update business hours for LA Trip Map locations using web search.
This script demonstrates the process of finding hours via web search instead of Google Places API.
"""

import json
import requests
import re
from datetime import datetime

SUPABASE_URL = 'https://jgvckilmltimabfdvaly.supabase.co'
SUPABASE_KEY = 'sb_publishable_TiO_hACvlsn4mfFazsHTlg_i9BnzCfr'

headers = {
    'apikey': SUPABASE_KEY,
    'Authorization': f'Bearer {SUPABASE_KEY}',
    'Content-Type': 'application/json',
    'Prefer': 'return=representation'
}

def fetch_locations():
    """Fetch all locations from Supabase"""
    try:
        response = requests.get(
            f'{SUPABASE_URL}/rest/v1/locations',
            headers=headers,
            params={'select': 'id,name,address,hours'}
        )
        response.raise_for_status()
        return response.json()
    except Exception as e:
        print(f'Error fetching locations: {e}')
        return None

def update_location_hours(location_id, hours_data):
    """Update a location's hours in Supabase"""
    try:
        response = requests.patch(
            f'{SUPABASE_URL}/rest/v1/locations',
            headers=headers,
            params={'id': f'eq.{location_id}'},
            json={'hours': hours_data}
        )
        response.raise_for_status()
        return True
    except Exception as e:
        print(f'Error updating location {location_id}: {e}')
        return False

def parse_hours_from_text(text):
    """
    Parse business hours from free-form text.
    Returns hours in Google Places API format.
    """
    hours_data = {
        'periods': [],
        'weekday_text': []
    }

    # Common patterns for hours
    # e.g., "Monday: 9:00 AM – 5:00 PM"
    # e.g., "Mon-Fri: 9am-5pm"
    # e.g., "Open 24 hours"

    day_map = {
        'monday': 1, 'mon': 1,
        'tuesday': 2, 'tue': 2, 'tues': 2,
        'wednesday': 3, 'wed': 3,
        'thursday': 4, 'thu': 4, 'thur': 4, 'thurs': 4,
        'friday': 5, 'fri': 5,
        'saturday': 6, 'sat': 6,
        'sunday': 0, 'sun': 0
    }

    # Check for 24 hours
    if '24 hour' in text.lower() or 'open 24' in text.lower():
        hours_data['weekday_text'] = ['Open 24 hours']
        # Single period with no close time indicates 24/7
        hours_data['periods'] = [{'open': {'day': 0, 'time': '0000'}}]
        return hours_data

    lines = text.split('\n')
    for line in lines:
        # Try to find day and time patterns
        # Pattern: "Monday: 9:00 AM – 5:00 PM" or "Monday 9:00 AM – 5:00 PM"
        match = re.search(r'(monday|tuesday|wednesday|thursday|friday|saturday|sunday|mon|tue|wed|thu|fri|sat|sun)[:\s]+(\d{1,2}):?(\d{2})?\s*(am|pm)?.*?(\d{1,2}):?(\d{2})?\s*(am|pm)?', line, re.IGNORECASE)

        if match:
            day_str = match.group(1).lower()
            day_num = day_map.get(day_str, None)

            if day_num is not None:
                # Parse open time
                open_hour = int(match.group(2))
                open_min = int(match.group(3)) if match.group(3) else 0
                open_period = match.group(4).lower() if match.group(4) else 'am'

                # Convert to 24-hour format
                if open_period == 'pm' and open_hour != 12:
                    open_hour += 12
                elif open_period == 'am' and open_hour == 12:
                    open_hour = 0

                open_time = f'{open_hour:02d}{open_min:02d}'

                # Parse close time if present
                if match.group(5):
                    close_hour = int(match.group(5))
                    close_min = int(match.group(6)) if match.group(6) else 0
                    close_period = match.group(7).lower() if match.group(7) else 'pm'

                    if close_period == 'pm' and close_hour != 12:
                        close_hour += 12
                    elif close_period == 'am' and close_hour == 12:
                        close_hour = 0

                    close_time = f'{close_hour:02d}{close_min:02d}'

                    period = {
                        'open': {'day': day_num, 'time': open_time},
                        'close': {'day': day_num, 'time': close_time}
                    }
                    hours_data['periods'].append(period)
                    hours_data['weekday_text'].append(line.strip())

    return hours_data if hours_data['periods'] else None

def save_to_json(data, filename):
    """Save data to JSON file"""
    with open(filename, 'w') as f:
        json.dump(data, f, indent=2)
    print(f'Saved to {filename}')

def main():
    import sys

    command = sys.argv[1] if len(sys.argv) > 1 else 'fetch'

    if command == 'fetch':
        print('Fetching locations from Supabase...')
        locations = fetch_locations()

        if locations:
            print(f'Found {len(locations)} locations\n')

            # Save locations to file
            save_to_json(locations, 'locations.json')

            # Show locations without hours
            without_hours = [loc for loc in locations if not loc.get('hours')]
            print(f'\nLocations without hours: {len(without_hours)}')

            for loc in without_hours:
                print(f"  - {loc['name']} ({loc.get('address', 'no address')})")

            print('\nNext: Use web search to find hours for each location')
            print('Then run: python3 fetch_hours.py update hours-data.json')

    elif command == 'update':
        if len(sys.argv) < 3:
            print('Usage: python3 fetch_hours.py update hours-data.json')
            return

        hours_file = sys.argv[2]

        print(f'Loading hours data from {hours_file}...')
        with open(hours_file, 'r') as f:
            hours_data = json.load(f)

        print(f'Found hours data for {len(hours_data)} locations\n')

        success_count = 0
        fail_count = 0

        for item in hours_data:
            location_id = item['id']
            hours = item.get('hours')
            name = item.get('name', location_id)

            if not hours:
                print(f'Skipping {name} (no hours data)')
                continue

            print(f'Updating {name}...')
            if update_location_hours(location_id, hours):
                print(f'  ✓ Success')
                success_count += 1
            else:
                print(f'  ✗ Failed')
                fail_count += 1

        print(f'\n📊 Summary:')
        print(f'  ✅ Success: {success_count}')
        print(f'  ❌ Failed: {fail_count}')

    elif command == 'check':
        print('Checking current hours data in database...')
        locations = fetch_locations()

        if locations:
            with_hours = [loc for loc in locations if loc.get('hours')]
            without_hours = [loc for loc in locations if not loc.get('hours')]

            print(f'\nTotal locations: {len(locations)}')
            print(f'With hours: {len(with_hours)}')
            print(f'Without hours: {len(without_hours)}')

            if without_hours:
                print('\nLocations without hours:')
                for loc in without_hours:
                    print(f"  - {loc['name']} ({loc.get('address', 'no address')})")

    else:
        print('Usage:')
        print('  python3 fetch_hours.py fetch          # Fetch locations')
        print('  python3 fetch_hours.py update FILE    # Update with hours data')
        print('  python3 fetch_hours.py check          # Check status')

if __name__ == '__main__':
    main()
